src/wp-includes/canonical.php

```diff
@@ -740,6 +754,30 @@ function lowercase_octets( $matches ) {
 		$requested_url = preg_replace_callback( '|%[a-fA-F0-9][a-fA-F0-9]|', 'lowercase_octets', $requested_url );
 	}
 
+	if (
+		$redirect_obj &&
+		is_a( $redirect_obj, 'WP_Post' )
+	) {
+		$post_status_obj = get_post_status_object( get_post_status( $redirect_obj ) );
+		if (
+			// Never redirect for unregistered post statuses.
+			! $post_status_obj ||
+			// Never redirect internal or protected post statuses.
+			$post_status_obj->internal ||
+			$post_status_obj->protected ||
```

Comment:
Why is this logic different from `is_post_status_viewable`?

Reply:
It accounts for logged in users having `read_post` caps on private statuses. Internal (trash, drafts) and protected (future) are not readable on the front end. 

Strictly speaking it should be `is_post_status_publicly_viewable()` but I went with consistency with `is_post_type_viewable()`.